src/services/pcn-bridge/src/Bridge_dp.c

```diff
@@ -0,0 +1,363 @@
+#include <uapi/linux/if_vlan.h>
+
+#include <uapi/linux/in.h>
+#include <uapi/linux/ip.h>
+#include <uapi/linux/pkt_cls.h>
+
+enum { BPDU_REASON = 1, BROADCAST_REASON };
+
+enum { PORT_MODE_ACCESS = 0, PORT_MODE_TRUNK };
+
+enum entry_type { STATIC = 0, DYNAMIC };
+
+#define VLAN_WILDCARD 0x00
+
+#if STP_ENABLED
+enum stp_state {
+  STP_DISABLED = 1 << 0,   /* 8.4.5: See note above. */
+  STP_LISTENING = 1 << 1,  /* 8.4.2: Not learning or relaying frames. */
+  STP_LEARNING = 1 << 2,   /* 8.4.3: Learning but not relaying frames. */
+  STP_FORWARDING = 1 << 3, /* 8.4.4: Learning and relaying frames. */
+  STP_BLOCKING = 1 << 4    /* 8.4.1: Initial boot state. */
+};
+
+/* Returns true if 'state' is one in which packets received on a port should
+ * be forwarded, false otherwise.
+ *
+ * Returns true if 'state' is STP_DISABLED, since presumably in that case the
+ * port should still work, just not have STP applied to it. */
+static __always_inline u32 stp_forward_in_state(u32 state) {
+  return (state & (STP_DISABLED | STP_FORWARDING)) != 0;
+}
+
+/* Returns true if 'state' is one in which MAC learning should be done on
+ * packets received on a port, false otherwise.
+ *
+ * Returns true if 'state' is STP_DISABLED, since presumably in that case the
+ * port should still work, just not have STP applied to it. */
+static __always_inline u32 stp_learn_in_state(u32 state) {
+  return (state & (STP_DISABLED | STP_LEARNING | STP_FORWARDING)) != 0;
+}
+#endif
+
+/*
+ * Allowed Vlans table is used to set and verify vlans associated to a port
+ * in access or trunk mode. The usage is provided in the following example.
+ */
+struct port_vlan_key {
+  u32 port;
+  u16 vlan;
+} __attribute__((packed));
+
+struct port_vlan_value {
+  u16 vlan;
+  u16 stp_state;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct port_vlan_key, struct port_vlan_value, port_vlan,
+          1024);
+/*
+ * port_vlan usage example:
+ * <port> <vlan> -> vlanid
+ *
+ * port 1 access vlan=10
+ * 1 VLAN_WILDCARD -> 10
+ *
+ * port 3 trunk vlan=30,40
+ * 3 30 -> 30
+ * 3 40 -> 40
+ *
+ * port 4 trunk vlan=all
+ * 4 VLAN_WILDCARD -> VLAN_WILDCARD
+ */
+
+/*
+ * Fowarding Table: VLAN + MAC Address to port association.  This table is
+ * filled
+ * in the learning phase and then is used in the forwarding phase to decide
+ * where to send a packet
+ */
+struct fwd_key {
+  u16 vlan;
+  __be64 mac;
+} __attribute__((packed));
+
+struct fwd_entry {
+  u32 timestamp;
+  u32 port;
+  enum entry_type type;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct fwd_key, struct fwd_entry, fwdtable, 1024);
+BPF_TABLE("array", int, uint32_t, timestamp, 1);
+
+static __always_inline u32 time_get_sec() {
+  int key = 0;
+  u32 *ts = timestamp.lookup(&key);
+  if (ts)
+    return *ts;
+
+  return 0;
+}
+
+/*
+ * Port Table: Saves the state of the ports
+ * TODO: what is the best option in this case, hash vs array?
+ */
+struct port {
+  u16 mode;
+  u16 native_vlan;
+  bool native_vlan_enabled;
```

Comment:
Why is the "native_vlan" here? I thought "native vlan" is an option of the entire bridge, not an option of each single port.

Reply:
No, from the start the datamodel had the "native-vlan" leaf at "root->ports->trunk" level, even in the old bridge was the same.

Reply:
Got it. Checked, also Cisco has per-port config of native VLANs. So, let's keep as it is. thanks!.